stl/inc/xutility

```diff
@@ -4963,35 +4958,80 @@ _NODISCARD bool lexicographical_compare(
+            _Strong_order::_Cpo, void>;
+};
+
+template <class _Elem1, class _Elem2>
+struct _Lex_compare_three_way_memcmp_classify_comp<_Elem1, _Elem2, _Weak_order::_Cpo> {
+    using _Comp =
+        conditional_t<_Lex_compare_memcmp_classify_elements<_Elem1, _Elem2> && _Can_weak_order<_Elem1, _Elem2>,
+            _Weak_order::_Cpo, void>;
+};
+
+template <class _Elem1, class _Elem2>
+struct _Lex_compare_three_way_memcmp_classify_comp<_Elem1, _Elem2, _Partial_order::_Cpo> {
+    using _Comp =
+        conditional_t<_Lex_compare_memcmp_classify_elements<_Elem1, _Elem2> && _Can_partial_order<_Elem1, _Elem2>,
+            _Partial_order::_Cpo, void>;
+};
+
+template <class _It1, class _It2, class _Comp>
+using _Lex_compare_three_way_memcmp_classify =
+    conditional_t<_Iterators_are_contiguous<_It1, _It2> && !_Iterator_is_volatile<_It1> && !_Iterator_is_volatile<_It2>,
+        typename _Lex_compare_three_way_memcmp_classify_comp<_Iter_value_t<_It1>, _Iter_value_t<_It2>, _Comp>::_Comp,
+        void>;
+
 template <class _InIt1, class _InIt2, class _Cmp>
-_NODISCARD constexpr auto lexicographical_compare_three_way(
-    _InIt1 _First1, _InIt1 _Last1, _InIt2 _First2, _InIt2 _Last2, _Cmp _Comp) -> decltype(_Comp(*_First1, *_First2)) {
+_NODISCARD constexpr auto lexicographical_compare_three_way(const _InIt1 _First1, const _InIt1 _Last1,
+    const _InIt2 _First2, const _InIt2 _Last2, _Cmp _Comp) -> decltype(_Comp(*_First1, *_First2)) {
     _Adl_verify_range(_First1, _Last1);
     _Adl_verify_range(_First2, _Last2);
     auto _UFirst1      = _Get_unwrapped(_First1);
     const auto _ULast1 = _Get_unwrapped(_Last1);
     auto _UFirst2      = _Get_unwrapped(_First2);
     const auto _ULast2 = _Get_unwrapped(_Last2);
-    using _UIt1        = decltype(_UFirst1);
-    using _UIt2        = decltype(_UFirst2);
-    using _Ty1         = remove_const_t<remove_pointer_t<_UIt1>>;
-    using _Ty2         = remove_const_t<remove_pointer_t<_UIt2>>;
-
-    if constexpr (
-        conjunction_v<is_same<_Cmp, compare_three_way>, bool_constant<_Iterators_are_contiguous<_UIt1, _UIt2>>,
-            disjunction<
-#ifdef __cpp_lib_byte
-                conjunction<is_same<_Ty1, byte>, is_same<_Ty2, byte>>,
-#endif // __cpp_lib_byte
-                conjunction<_Is_character<_Ty1>, is_unsigned<_Ty1>, _Is_character<_Ty2>, is_unsigned<_Ty2>>>>) {
+
+    using _Memcmp_pred = _Lex_compare_three_way_memcmp_classify<decltype(_UFirst1), decltype(_UFirst2), _Cmp>;
+    if constexpr (!is_void_v<_Memcmp_pred>) {
         if (!_STD is_constant_evaluated()) {
             const auto _Num1 = static_cast<size_t>(_ULast1 - _UFirst1);
             const auto _Num2 = static_cast<size_t>(_ULast2 - _UFirst2);
             const int _Ans   = _Memcmp_count(_UFirst1, _UFirst2, (_STD min) (_Num1, _Num2));
             if (_Ans == 0) {
-                return _Num1 <=> _Num2;
+                return _Memcmp_pred{}(_Num1, _Num2);

```

Comment:
WG21-N4892 25.8.12 [alg.three.way]/3 "*Returns:* [...] or `(e1 - b1) <=> (e2 - b2)` if no such integer exists."

After some thought, I don't believe this affects correctness, but for clarity I believe we should directly `<=>` compare the lengths here, instead of using the `_Memcmp_pred` returned by the metaprogramming. This will match the Standardese, and the classic `lexicographical_compare` implementation more closely. (For the classic impl, this is critical, as the metaprogrammed machinery is capable of reversing the ordering, which we don't want to do for the lengths of equal-prefix sequences. For the three-way impl, this is cosmetic, as we activate the optimization only for strong_order/weak_order/partial_order, none of which reverse things like std::greater do).